moveit_ros/perception/pointcloud_octomap_updater/src/pointcloud_octomap_updater.cpp

```diff
@@ -297,39 +296,50 @@ void PointCloudOctomapUpdater::cloudMsgCallback(const sensor_msgs::PointCloud2::
       }
     }
 
-    /* compute the free cells along each ray that ends at an occupied cell */
-    for (const octomap::OcTreeKey& occupied_cell : occupied_cells)
-      if (tree_->computeRayKeys(sensor_origin, tree_->keyToCoord(occupied_cell), key_ray_))
-        free_cells.insert(key_ray_.begin(), key_ray_.end());
-
-    /* compute the free cells along each ray that ends at a model cell */
-    for (const octomap::OcTreeKey& model_cell : model_cells)
-      if (tree_->computeRayKeys(sensor_origin, tree_->keyToCoord(model_cell), key_ray_))
-        free_cells.insert(key_ray_.begin(), key_ray_.end());
-
-    /* compute the free cells along each ray that ends at a clipped cell */
-    for (const octomap::OcTreeKey& clip_cell : clip_cells)
-      if (tree_->computeRayKeys(sensor_origin, tree_->keyToCoord(clip_cell), key_ray_))
-        free_cells.insert(key_ray_.begin(), key_ray_.end());
+    if (update_method == UpdateMethod::INCREMENTAL)
+    {
+      /* If we are performing an incremental update, do ray tracing to find which cells this point cloud indicates
+       * should
+       * be free */
+
+      /* compute the free cells along each ray that ends at an occupied cell */
+      for (const octomap::OcTreeKey& occupied_cell : occupied_cells)
+        if (tree_->computeRayKeys(sensor_origin, tree_->keyToCoord(occupied_cell), key_ray_))
+          free_cells.insert(key_ray_.begin(), key_ray_.end());
+
+      /* compute the free cells along each ray that ends at a model cell */
+      for (const octomap::OcTreeKey& model_cell : model_cells)
+        if (tree_->computeRayKeys(sensor_origin, tree_->keyToCoord(model_cell), key_ray_))
+          free_cells.insert(key_ray_.begin(), key_ray_.end());
+
+      /* compute the free cells along each ray that ends at a clipped cell */
+      for (const octomap::OcTreeKey& clip_cell : clip_cells)
+        if (tree_->computeRayKeys(sensor_origin, tree_->keyToCoord(clip_cell), key_ray_))
+          free_cells.insert(key_ray_.begin(), key_ray_.end());
+    }
   }
   catch (...)
   {
     tree_->unlockRead();
-    return;
+    ROS_ERROR_NAMED(LOGNAME, "Internal error while updating octree");
+    return false;
   }
 
   tree_->unlockRead();
 
-  /* cells that overlap with the model are not occupied */
-  for (const octomap::OcTreeKey& model_cell : model_cells)
-    occupied_cells.erase(model_cell);
-
-  /* occupied cells are not free */
-  for (const octomap::OcTreeKey& occupied_cell : occupied_cells)
-    free_cells.erase(occupied_cell);
+  if (update_method == UpdateMethod::INCREMENTAL)
+  {
+    /* occupied cells are not free */
```

Comment:
Doesn't this fit in the block above where `occupied_cells` is already being iterated?

Reply:
I believe I separated it out because this loop is only modifying local variables so it can be performed without locking the OcTree

Reply:
Oh I think I see what you mean.  This loop existed separately in the original so I just brought it over.  You mean this loop 
```     
for (const octomap::OcTreeKey& occupied_cell : occupied_cells)
    if (tree_->computeRayKeys(sensor_origin, tree_->keyToCoord(occupied_cell), key_ray_))
        free_cells.insert(key_ray_.begin(), key_ray_.end());
```
right?  It's possible that these could be combined.  I think combining them might have unexpected consequences if there are somehow multiple occupied cells on the same ray, whereas by performing this separately we guarantee none of `occupied_cells` accidentally get marked free.  But I don't know if that's something we really need to worry about or not.

Reply:
I see, makes sense!